### PR TITLE
LUCENE-9878: enable redundantNullCheck in ecjLint

### PR DESCRIPTION
**Pull request by rmuir**
Detects common cases of unreachable/dead code.

Don't be confused by the name, this isn't enabling annotation-based null analyzer or discouraging safety checks in the code.

This is simple flow analysis to detect cases like the following:

```
foo = null;
if (foo != null) {
  // provably dead code
}
```

```
foo.invokeMethod();
if (foo != null) {
  // provably dead code, would have thrown exc
}
```

I reviewed all the errors, and the check seems productive. I think a lot of this stuff happens as the code "ages" over time. Lots of WTFs  It even found what appears to be a seriously broken `equals()` implementation.

For generated javacc code, the check is disabled via `SuppressWarnings("unused")` because javacc generates strange/bad code such as:

```
if ("" == null)
```

For TestStressNRTReplication's `startNode()` method, the check is also disabled because analysis folds the "test evilness controls" which are static final constants. This itself is a WTF, shouldn't we instead randomize these evil things in our tests rather than hardcoding them to specific values?

### CODE REVIEW

File: lucene/grouping/src/java/org/apache/lucene/search/grouping/SearchGroup.java [hunk 1]
```diff
@@ -143,9 +143,9 @@ public boolean equals(Object _other) {
       if (_other instanceof MergedGroup) {
         MergedGroup<?> other = (MergedGroup<?>) _other;
         if (groupValue == null) {
-          return other == null;
+          return other.groupValue == null;
```
Comment (rmuir) [inline lucene/grouping/src/java/org/apache/lucene/search/grouping/SearchGroup.java:146]: I think this stuff was a real bug, causing `equals` to basically always return false. I don't know how it is used by this thing (caching or performance or whatever). Detector knows that `other == _other` and that `_other instanceof MergedGroup`, hence it simply cannot be null here.

But the bigger issue is that we should be comparing against `other.groupValue` in this method.

Comment (mikemccand) [reply lucene/grouping/src/java/org/apache/lucene/search/grouping/SearchGroup.java:146]: Wow, this is an awesome catch and shows how useful static analysis is!!

**Review by mikemccand — APPROVED** [review]
This looks great @rmuir!  I'll try to merge shortly.

File: lucene/core/src/java/org/apache/lucene/index/CheckIndex.java [hunk 2]
```diff
@@ -652,14 +651,6 @@ public Status checkIndex(List<String> onlySegments) throws IOException {
         Sort indexSort = info.info.getIndexSort();
         if (indexSort != null) {
           msg(infoStream, "    sort=" + indexSort);
-          if (previousIndexSort != null) {
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/index/CheckIndex.java:655]: This is indeed dead code, but that's crazy!  I think the intention here was to set `previousIndexSort = null` OUTSIDE the `for` loop, but that didn't happen, so this code is clearly dead.

**Pull request merged**